storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java

```diff
@@ -64,302 +64,395 @@
-    private final KafkaConsumer<byte[], byte[]> consumer;
-    private final String metadataTopicName;
+    private final Consumer<byte[], byte[]> consumer;
     private final RemotePartitionMetadataEventHandler remotePartitionMetadataEventHandler;
     private final RemoteLogMetadataTopicPartitioner topicPartitioner;
     private final Time time;
 
-    // It indicates whether the closing process has been started or not. If it is set as true,
-    // consumer will stop consuming messages, and it will not allow partition assignments to be updated.
-    private volatile boolean closing = false;
-
-    // It indicates whether the consumer needs to assign the partitions or not. This is set when it is
-    // determined that the consumer needs to be assigned with the updated partitions.
-    private volatile boolean assignPartitions = false;
+    // It indicates whether the ConsumerTask is closed or not.
+    private volatile boolean isClosed = false;
+    // It indicates whether the user topic partition assignment to the consumer has changed or not. If the assignment
+    // has changed, the consumer will eventually start tracking the newly assigned partitions and stop tracking the
+    // ones it is no longer assigned to.
+    private volatile boolean hasAssignmentChanged = true;
 
     // It represents a lock for any operations related to the assignedTopicPartitions.
     private final Object assignPartitionsLock = new Object();
 
     // Remote log metadata topic partitions that consumer is assigned to.
-    private volatile Set<Integer> assignedMetaPartitions = Collections.emptySet();
+    private volatile Set<Integer> assignedMetadataPartitions = Collections.emptySet();
 
     // User topic partitions that this broker is a leader/follower for.
-    private Set<TopicIdPartition> assignedTopicPartitions = Collections.emptySet();
+    private volatile Map<TopicIdPartition, UserTopicIdPartition> assignedUserTopicIdPartitions = Collections.emptyMap();
+    private volatile Set<TopicIdPartition> processedAssignmentOfUserTopicIdPartitions = Collections.emptySet();
 
-    // Map of remote log metadata topic partition to consumed offsets. Received consumer records
-    // may or may not have been processed based on the assigned topic partitions.
-    private final Map<Integer, Long> partitionToConsumedOffsets = new ConcurrentHashMap<>();
+    private long uninitializedAt;
+    private boolean isAllUserTopicPartitionsInitialized;
 
-    // Map of remote log metadata topic partition to processed offsets that were synced in committedOffsetsFile.
-    private Map<Integer, Long> lastSyncedPartitionToConsumedOffsets = Collections.emptyMap();
+    // Map of remote log metadata topic partition to consumed offsets.
+    private final Map<Integer, Long> readOffsetsByMetadataPartition = new ConcurrentHashMap<>();
+    private final Map<TopicIdPartition, Long> readOffsetsByUserTopicPartition = new HashMap<>();
```

Comment:
nit: reuse `latest` concept
```suggestion
    private final Map<Integer, Long> latestOffsetsByMetadataPartition = new ConcurrentHashMap<>();
    private final Map<TopicIdPartition, Long> latestOffsetsByUserTopicPartition = new HashMap<>();
```

Reply:
what is the reasoning to rename readOffsets to latestOffsets? 

`readOffsets` denotes the offsets so far read by the consumer.

Reply:
We used "read" and "received" which is a bit confusing. maybe `latestRead*` and align naming would be better?

Reply:
Can we rename `received` to `read` instead? `latestOffset` denotes log-end-offset which will be confusing.

Reply:
Sure, if @abhijeetk88 agrees, let's go with this option

Reply:
Done